运行报错了INTERNAL ASSERT FAILED怎么解决啊？

在运行：`wavs = chat.infer(texts, use_decoder=True)` 时报了如下错误：

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[5], [line 1](vscode-notebook-cell:?execution_count=5&line=1)
----> [1](vscode-notebook-cell:?execution_count=5&line=1) wavs = chat.infer(texts, use_decoder=True)

File ~/git/ChatTTS/ChatTTS/core.py:153, in Chat.infer(self, text, skip_refine_text, refine_text_only, params_refine_text, params_infer_code, use_decoder)
    [150](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:150) else:
    [151](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:151)     mel_spec = [self.pretrain_models['dvae'](i[None].permute(0,2,1)) for i in result['ids']]
--> [153](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:153) wav = [self.pretrain_models['vocos'].decode(i).cpu().numpy() for i in mel_spec]
    [155](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:155) return wav

File ~/git/ChatTTS/ChatTTS/core.py:153, in <listcomp>(.0)
    [150](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:150) else:
    [151](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:151)     mel_spec = [self.pretrain_models['dvae'](i[None].permute(0,2,1)) for i in result['ids']]
--> [153](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:153) wav = [self.pretrain_models['vocos'].decode(i).cpu().numpy() for i in mel_spec]
    [155](https://file+.vscode-resource.vscode-cdn.net/Users/zhaohui/git/ChatTTS/~/git/ChatTTS/ChatTTS/core.py:155) return wav

File /usr/local/lib/python3.11/site-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    [112](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/torch/utils/_contextlib.py:112) @functools.wraps(func)
    [113](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/torch/utils/_contextlib.py:113) def decorate_context(*args, **kwargs):
    [114](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/torch/utils/_contextlib.py:114)     with ctx_factory():
--> [115](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/torch/utils/_contextlib.py:115)         return func(*args, **kwargs)

File /usr/local/lib/python3.11/site-packages/vocos/pretrained.py:113, in Vocos.decode(self, features_input, **kwargs)
    [101](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/vocos/pretrained.py:101) """
...
---> [67](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/vocos/heads.py:67) S = mag * (x + 1j * y)
     [68](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/vocos/heads.py:68) audio = self.istft(S)
     [69](https://file+.vscode-resource.vscode-cdn.net/usr/local/lib/python3.11/site-packages/vocos/heads.py:69) return audio
RuntimeError: c10::isComplexType(input.scalar_type()) && c10::isComplexType(other.scalar_type()) INTERNAL ASSERT FAILED at "/Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/native/mps/operations/BinaryKernel.mm":340, please report a bug to PyTorch. 
```

此外在运行期间一直有warning
```
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] Traceback (most recent call last):
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 727, in _convert_frame
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     result = inner_convert(frame, cache_entry, hooks, frame_state)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 383, in _convert_frame_assert
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_product = _compile(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                        ^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 646, in _compile
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     guarded_code = compile_inner(code, one_graph, hooks, transform)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     r = func(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 562, in compile_inner
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     out_code = transform_code_object(code, transform)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/bytecode_transformation.py", line 1033, in transform_code_object
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     transformations(instructions, code_options)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 151, in _fn
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return fn(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/convert_frame.py", line 527, in transform
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     tracer.run()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 2128, in run
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     super().run()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 818, in run
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     and self.step()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 781, in step
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     getattr(self, inst.opname)(inst)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/symbolic_convert.py", line 2243, in RETURN_VALUE
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     self.output.compile_subgraph(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/output_graph.py", line 919, in compile_subgraph
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     self.compile_and_call_fx_graph(tx, list(reversed(stack_values)), root)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 81, in inner
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return func(*args, **kwds)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/output_graph.py", line 1087, in compile_and_call_fx_graph
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fn = self.call_user_compiler(gm)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     r = func(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/output_graph.py", line 1159, in call_user_compiler
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/output_graph.py", line 1140, in call_user_compiler
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fn = compiler_fn(gm, self.example_inputs())
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/repro/after_dynamo.py", line 117, in debug_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_gm = compiler_fn(gm, example_inputs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/__init__.py", line 1668, in __call__
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return compile_fx(model_, inputs_, config_patches=self.config)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 1168, in compile_fx
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return aot_autograd(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/backends/common.py", line 55, in compiler_fn
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     cg = aot_module_simplified(gm, example_inputs, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 887, in aot_module_simplified
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fn = create_aot_dispatcher_function(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     r = func(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_functorch/aot_autograd.py", line 600, in create_aot_dispatcher_function
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 425, in aot_wrapper_dedupe
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/runtime_wrappers.py", line 630, in aot_wrapper_synthetic_base
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_functorch/_aot_autograd/jit_compile_runtime_wrappers.py", line 97, in aot_dispatch_base
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fw = compiler(fw_module, updated_flat_args)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     r = func(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 1100, in fw_compiler_base
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return inner_compile(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/repro/after_aot.py", line 83, in debug_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     inner_compiled_fn = compiler_fn(gm, example_inputs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/debug.py", line 305, in inner
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return fn(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/Cellar/python@3.11/3.11.4_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 81, in inner
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return func(*args, **kwds)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 320, in compile_fx_inner
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_graph = fx_codegen_and_compile(
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                      ^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/compile_fx.py", line 550, in fx_codegen_and_compile
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     compiled_fn = graph.compile_to_fn()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                   ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 1116, in compile_to_fn
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     return self.compile_to_module().call
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_dynamo/utils.py", line 244, in time_wrapper
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     r = func(*args, **kwargs)
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]         ^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 1066, in compile_to_module
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]                                                              ^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 1039, in codegen
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     self.init_wrapper_code()
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]   File "/usr/local/lib/python3.11/site-packages/torch/_inductor/graph.py", line 993, in init_wrapper_code
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]     assert wrapper_code_gen_cls is not None, f"Device {device_type} not supported"
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] AssertionError: Device mps not supported
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] 
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
[2024-05-30 21:21:57,649] torch._dynamo.convert_frame: [WARNING] 
```

基本信息
* pytorch：2.2.2
* Python: 3.11.4
* 设备：Apple M3 Pro
* OS：14.5 (23F79)

--- a/ChatTTS/core.py
+++ b/ChatTTS/core.py
@@ -80,7 +80,10 @@ class Chat:
             self.logger.log(logging.INFO, f'use {device}')
             
         if vocos_config_path:
-            vocos = Vocos.from_hparams(vocos_config_path).to(device).eval()
+            vocos = Vocos.from_hparams(vocos_config_path).to(
+                # vocos on mps will crash, use cpu fallback
+                "cpu" if torch.backends.mps.is_available() else device
+            ).eval()
             assert vocos_ckpt_path, 'vocos_ckpt_path should not be None'
             vocos.load_state_dict(torch.load(vocos_ckpt_path))
             self.pretrain_models['vocos'] = vocos
@@ -90,7 +93,7 @@ class Chat:
             cfg = OmegaConf.load(dvae_config_path)
             dvae = DVAE(**cfg).to(device).eval()
             assert dvae_ckpt_path, 'dvae_ckpt_path should not be None'
-            dvae.load_state_dict(torch.load(dvae_ckpt_path, map_location='cpu'))
+            dvae.load_state_dict(torch.load(dvae_ckpt_path))
             self.pretrain_models['dvae'] = dvae
             self.logger.log(logging.INFO, 'dvae loaded.')
             
@@ -98,7 +101,7 @@ class Chat:
             cfg = OmegaConf.load(gpt_config_path)
             gpt = GPT_warpper(**cfg).to(device).eval()
             assert gpt_ckpt_path, 'gpt_ckpt_path should not be None'
-            gpt.load_state_dict(torch.load(gpt_ckpt_path, map_location='cpu'))
+            gpt.load_state_dict(torch.load(gpt_ckpt_path))
             if compile and 'cuda' in str(device):
                 gpt.gpt.forward = torch.compile(gpt.gpt.forward,  backend='inductor', dynamic=True)
             self.pretrain_models['gpt'] = gpt
@@ -139,21 +142,21 @@ class Chat:
         
         if not isinstance(text, list): 
             text = [text]
-        
+
         if do_text_normalization:
             for i, t in enumerate(text):
                 _lang = detect_language(t) if lang is None else lang
-                self.init_normalizer(_lang)
-                text[i] = self.normalizer[_lang](t)
-                if _lang == 'zh':
-                    text[i] = apply_half2full_map(text[i])
-            
+                if self.init_normalizer(_lang):
+                    text[i] = self.normalizer[_lang](t)
+                    if _lang == 'zh':
+                        text[i] = apply_half2full_map(text[i])
+
         for i, t in enumerate(text):
             invalid_characters = count_invalid_characters(t)
             if len(invalid_characters):
                 self.logger.log(logging.WARNING, f'Invalid characters found! : {invalid_characters}')
                 text[i] = apply_character_map(t)
-                
+
         if not skip_refine_text:
             text_tokens = refine_text(self.pretrain_models, text, **params_refine_text)['ids']
             text_tokens = [i[i < self.pretrain_models['tokenizer'].convert_tokens_to_ids('[break_0]')] for i in text_tokens]
@@ -170,7 +173,9 @@ class Chat:
         else:
             mel_spec = [self.pretrain_models['dvae'](i[None].permute(0,2,1)) for i in result['ids']]
             
-        wav = [self.pretrain_models['vocos'].decode(i).cpu().numpy() for i in mel_spec]
+        wav = [self.pretrain_models['vocos'].decode(
+            i.cpu() if torch.backends.mps.is_available() else i
+        ).cpu().numpy() for i in mel_spec]
         
         return wav
     
@@ -180,21 +185,37 @@ class Chat:
         std, mean = self.pretrain_models['spk_stat'].chunk(2)
         return torch.randn(dim, device=std.device) * std + mean
     
-    def init_normalizer(self, lang):
-        
-        if lang not in self.normalizer:
-            if lang == 'zh':
-                try:
-                    from tn.chinese.normalizer import Normalizer
-                except:
-                    self.logger.log(logging.WARNING, f'Package WeTextProcessing not found! \
-                        Run: conda install -c conda-forge pynini=2.1.5 && pip install WeTextProcessing')
-                self.normalizer[lang] = Normalizer().normalize
-            else:
-                try:
-                    from nemo_text_processing.text_normalization.normalize import Normalizer
-                except:
-                    self.logger.log(logging.WARNING, f'Package nemo_text_processing not found! \
-                        Run: conda install -c conda-forge pynini=2.1.5 && pip install nemo_text_processing')
-                self.normalizer[lang] = partial(Normalizer(input_case='cased', lang=lang).normalize, verbose=False, punct_post_process=True)
+    def init_normalizer(self, lang) -> bool:
 
+        if lang in self.normalizer:
+            return True
+
+        if lang == 'zh':
+            try:
+                from tn.chinese.normalizer import Normalizer
+                self.normalizer[lang] = Normalizer().normalize
+                return True
+            except:
+                self.logger.log(
+                    logging.WARNING,
+                    'Package WeTextProcessing not found!',
+                )
+                self.logger.log(
+                    logging.WARNING,
+                    'Run: conda install -c conda-forge pynini=2.1.5 && pip install WeTextProcessing',
+                )   
+        else:
+            try:
+                from nemo_text_processing.text_normalization.normalize import Normalizer
+                self.normalizer[lang] = partial(Normalizer(input_case='cased', lang=lang).normalize, verbose=False, punct_post_process=True)
+                return True
+            except:
+                self.logger.log(
+                    logging.WARNING,
+                    'Package nemo_text_processing not found!',
+                )
+                self.logger.log(
+                    logging.WARNING,
+                    'Run: conda install -c conda-forge pynini=2.1.5 && pip install nemo_text_processing',
+                )
+        return False

--- a/ChatTTS/model/gpt.py
+++ b/ChatTTS/model/gpt.py
@@ -184,11 +184,17 @@ class GPT_warpper(nn.Module):
             if attention_mask is not None:
                 attention_mask_cache[:, :attention_mask.shape[1]] = attention_mask
             
+            past_key_values = None
+
             for i in tqdm(range(max_new_token)):
         
-                model_input = self.prepare_inputs_for_generation(inputs_ids, 
-                    outputs.past_key_values if i!=0 else None, 
-                    attention_mask_cache[:, :inputs_ids.shape[1]], use_cache=True)
+                model_input = self.prepare_inputs_for_generation(
+                    inputs_ids, 
+                    past_key_values, 
+                    attention_mask_cache[:, :inputs_ids.shape[1]],
+                    use_cache=True,
+                )
+
             
                 if i == 0:
                     model_input['inputs_embeds'] = emb
@@ -201,8 +207,11 @@ class GPT_warpper(nn.Module):
                 
                 model_input['input_ids'] = None
                 outputs = self.gpt.forward(**model_input, output_attentions=return_attn)
+                del model_input
                 attentions.append(outputs.attentions)
                 hidden_states = outputs[0] # 🐻
+                past_key_values = outputs.past_key_values
+                del outputs
                 if return_hidden:
                     hiddens.append(hidden_states[:, -1])
 
@@ -227,23 +236,31 @@ class GPT_warpper(nn.Module):
                     
                 for logitsWarpers in LogitsWarpers:
                     logits = logitsWarpers(logits_token, logits)
+                del logits_token
                     
                 if i < min_new_token:
                     logits[:, eos_token] = -torch.inf
                 
                 scores = F.softmax(logits, dim=-1)
-            
+                del logits
+
                 idx_next = torch.multinomial(scores, num_samples=1)
                 
                 if not infer_text:
                     idx_next = rearrange(idx_next, "(b n) 1 -> b n", n=self.num_vq)
-                    finish = finish | (idx_next == eos_token).any(1)
+                    finish_or = (idx_next == eos_token).any(1)
+                    finish |= finish_or
+                    del finish_or
                     inputs_ids = torch.cat([inputs_ids, idx_next.unsqueeze(1)], 1)
                 else:
-                    finish = finish | (idx_next == eos_token).any(1)
+                    finish_or = (idx_next == eos_token).any(1)
+                    finish |= finish_or
+                    del finish_or
                     inputs_ids = torch.cat([inputs_ids, idx_next.unsqueeze(-1).expand(-1, -1, self.num_vq)], 1)
 
-                end_idx = end_idx + (~finish).int()
+                del idx_next
+
+                end_idx += (~finish).int().to(end_idx.device)
             
                 if finish.all():
                     break
@@ -256,7 +273,9 @@ class GPT_warpper(nn.Module):
                 hiddens = [hiddens[idx, :i] for idx, i in enumerate(end_idx.int())]
                     
             if not finish.all():
-                self.logger.warn(f'Incomplete result. hit max_new_token: {max_new_token}')    
+                self.logger.warn(f'Incomplete result. hit max_new_token: {max_new_token}')   
+
+            del finish 
                    
             return {
                 'ids': inputs_ids, 

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,3 +1,4 @@
+numpy<2.0.0
 omegaconf~=2.3.0
 torch~=2.1.0
 tqdm

